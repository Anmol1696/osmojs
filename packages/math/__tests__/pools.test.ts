import {
  calcMaxCoinsForPool,
  calcShareOutAmount,
  prettyPool,
} from "./../src/pool";
import priceResponse from "../../../__fixtures__/coingecko/api/v3/simple/price/data.json";
import poolResponse from "../../../__fixtures__/rpc/osmosis/gamm/v1beta1/pools/data.json";
import {
  noDecimals,
  calcPoolLiquidity,
  convertGammTokenToDollarValue,
  convertDollarValueToCoins,
  convertDollarValueToShares,
  calcCoinsNeededForValue,
  convertGeckoPricesToDenomPriceHash,
} from "../src/pool";
import cases from "jest-in-case";

const fakeBalances = [
  {
    denom: "gamm/pool/1",
    amount: "3971711067183317356",
  },
  {
    denom: "gamm/pool/497",
    amount: "349025380200104928",
  },
  {
    denom: "gamm/pool/604",
    amount: "35696235433901",
  },
  {
    denom: "gamm/pool/795",
    amount: "6240168637927540325",
  },
  {
    denom:
      "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
    amount: "838542",
  },
  {
    denom:
      "ibc/2DA9C149E9AD2BD27FEFA635458FB37093C256C1A940392634A16BEA45262604",
    amount: "2887086",
  },
  {
    denom:
      "ibc/46B44899322F3CD854D2D46DEEF881958467CDD4B3B10086DA49296BBED94BED",
    amount: "35409600",
  },
  {
    denom:
      "ibc/987C17B11ABC2B20019178ACE62929FE9840202CE79498E29FE8E5CB02B7C0A4",
    amount: "232663227",
  },
  {
    denom: "uosmo",
    amount: "5283728",
  },
];

export const omit = (obj, ...props) => {
  const result = { ...obj };
  props.forEach((prop) => {
    delete result[prop];
  });
  return result;
};

const pools = poolResponse.pools.map((p) => omit(p, "@type"));
const prices = convertGeckoPricesToDenomPriceHash(priceResponse);

describe("Test pool calculations", () => {
  cases(
    "calcPoolLiquidity",
    (opts) => {
      const pool = pools.find((pool) => pool.id === opts.poolId);
      const liquidity = calcPoolLiquidity(pool, prices);
      expect(noDecimals(liquidity)).toEqual(opts.liquidity);
    },
    [
      {
        name: "pool#1",
        poolId: "1",
        liquidity: "446862602",
      },
      {
        name: "pool#497",
        poolId: "497",
        liquidity: "65709813",
      },
      {
        name: "pool#604",
        poolId: "604",
        liquidity: "24905046",
      },
    ]
  );

  cases(
    "convertGammTokenToDollarValue",
    (opts) => {
      const value = convertGammTokenToDollarValue(opts.coin, pools, prices);
      expect(noDecimals(value)).toEqual(opts.value);
    },
    [
      { name: "gamm/pool/1", coin: fakeBalances[0], value: "5" },
      { name: "gamm/pool/497", coin: fakeBalances[1], value: "14" },
      { name: "gamm/pool/604", coin: fakeBalances[2], value: "6" },
    ]
  );

  cases(
    "convertDollarValueToCoins",
    (opts) => {
      const pool = pools.find(
        (pool) => pool.poolAssets.length === opts.poolLength
      );
      const coins = convertDollarValueToCoins(opts.value, pool, prices);
      expect(coins).toMatchSnapshot();
    },
    [
      {
        name: "2 token pool",
        value: 10,
        poolLength: 2,
      },
      {
        name: "3 token pool",
        value: 10,
        poolLength: 3,
      },
      {
        name: "4 token pool",
        value: 10,
        poolLength: 4,
      },
      {
        name: "0 dollar",
        value: 0,
        poolLength: 2,
      },
    ]
  );

  cases(
    "convertDollarValueToShares",
    (opts) => {
      const pool = pools.find((pool) => pool.id === opts.poolId);
      const shares = convertDollarValueToShares(opts.value, pool, prices);
      expect(noDecimals(shares)).toEqual(opts.shares);
    },
    [
      { name: "10 usd for pool#1", value: 10, poolId: "1", shares: "7" },
      { name: "100 usd for pool#497", value: 100, poolId: "497", shares: "2" },
      { name: "0 usd for pool#1", value: 0, poolId: "1", shares: "0" },
    ]
  );

  cases(
    "calcCoinsNeededForValue",
    (opts) => {
      const pool = pools.find((pool) => pool.id === opts.poolId);
      const prettiedPool = prettyPool(pool);
      const coinsNeeded = calcCoinsNeededForValue(
        prices,
        prettiedPool,
        opts.value
      );
      expect(coinsNeeded).toMatchSnapshot();
    },
    [
      { name: "10 for pool#1", poolId: "1", value: 10 },
      { name: "10 for pool#497", poolId: "497", value: 10 },
      { name: "10 for pool#604", poolId: "604", value: 10 },
      { name: "0 for pool#1", poolId: "1", value: 0 },
    ]
  );

  cases(
    "calcMaxCoinsForPool",
    (opts) => {
      const pool = pools.find((pool) => pool.id === opts.poolId);
      const prettiedPool = prettyPool(pool);
      const coinsNeeded = calcMaxCoinsForPool(
        prices,
        prettiedPool,
        fakeBalances
      );
      expect(coinsNeeded).toMatchSnapshot();
    },
    [
      { name: "pool#1", poolId: "1" },
      { name: "pool#604", poolId: "604" },
      { name: "pool#25", poolId: "25" },
    ]
  );

  cases(
    "calcShareOutAmount",
    (opts) => {
      const pool = pools.find((pool) => pool.id === opts.poolId);
      const prettiedPool = prettyPool(pool);
      const coinsNeeded = opts.useMaxCoins
        ? calcMaxCoinsForPool(prices, prettiedPool, fakeBalances)
        : calcCoinsNeededForValue(prices, prettiedPool, opts.value as number);
      const shareOutAmount = calcShareOutAmount(pool, coinsNeeded);
      expect(shareOutAmount).toEqual(opts.amount);
    },
    [
      {
        name: "pool#1",
        poolId: "1",
        useMaxCoins: true,
        amount: "44690927353631068872",
      },
      {
        name: "pool#604",
        poolId: "604",
        useMaxCoins: false,
        value: 10,
        amount: "51864273176244",
      },
      {
        name: "pool#497",
        poolId: "497",
        useMaxCoins: true,
        amount: "2222110943246410615",
      },
    ]
  );
});
