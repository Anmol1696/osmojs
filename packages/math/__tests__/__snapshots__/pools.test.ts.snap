// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test pool calculations calcCoinsNeededForValue 0 for pool#1 1`] = `
[
  {
    "amount": "0",
    "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
    "displayAmount": "0",
    "shareTotalValue": "0",
    "symbol": "ATOM",
    "totalDollarValue": "226964145.58592049",
    "unitRatio": "0",
  },
  {
    "amount": "0",
    "denom": "uosmo",
    "displayAmount": "0",
    "shareTotalValue": "0",
    "symbol": "OSMO",
    "totalDollarValue": "219898456.741054",
    "unitRatio": "0",
  },
]
`;

exports[`Test pool calculations calcCoinsNeededForValue 10 for pool#1 1`] = `
[
  {
    "amount": "136500",
    "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
    "displayAmount": "0.13650013650013650014",
    "shareTotalValue": "5",
    "symbol": "ATOM",
    "totalDollarValue": "226964145.58592049",
    "unitRatio": "2.202991131966e-8",
  },
  {
    "amount": "549450",
    "denom": "uosmo",
    "displayAmount": "0.54945054945054945055",
    "shareTotalValue": "5",
    "symbol": "OSMO",
    "totalDollarValue": "219898456.741054",
    "unitRatio": "2.273776757737e-8",
  },
]
`;

exports[`Test pool calculations calcCoinsNeededForValue 10 for pool#497 1`] = `
[
  {
    "amount": "275633",
    "denom": "ibc/46B44899322F3CD854D2D46DEEF881958467CDD4B3B10086DA49296BBED94BED",
    "displayAmount": "0.27563395810363836825",
    "shareTotalValue": "5",
    "symbol": "JUNO",
    "totalDollarValue": "33383106.20983148",
    "unitRatio": "1.4977635599792e-7",
  },
  {
    "amount": "549450",
    "denom": "uosmo",
    "displayAmount": "0.54945054945054945055",
    "shareTotalValue": "5",
    "symbol": "OSMO",
    "totalDollarValue": "32326707.4682188",
    "unitRatio": "1.5467087097922e-7",
  },
]
`;

exports[`Test pool calculations calcCoinsNeededForValue 10 for pool#604 1`] = `
[
  {
    "amount": "7051808",
    "denom": "ibc/987C17B11ABC2B20019178ACE62929FE9840202CE79498E29FE8E5CB02B7C0A4",
    "displayAmount": "7.05180822466496859125",
    "shareTotalValue": "5",
    "symbol": "STARS",
    "totalDollarValue": "13817947.976249658516",
    "unitRatio": "3.6184822873802e-7",
  },
  {
    "amount": "549450",
    "denom": "uosmo",
    "displayAmount": "0.54945054945054945055",
    "shareTotalValue": "5",
    "symbol": "OSMO",
    "totalDollarValue": "11087098.9624122",
    "unitRatio": "4.5097459822007e-7",
  },
]
`;

exports[`Test pool calculations calcMaxCoinsForPool pool#1 1`] = `
[
  {
    "amount": "838542",
    "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
  },
  {
    "amount": "3375361",
    "denom": "uosmo",
  },
]
`;

exports[`Test pool calculations calcMaxCoinsForPool pool#25 1`] = `
[
  {
    "amount": "0",
    "denom": "ibc/1DCC8A6CB5689018431323953344A9F6CC4D0BFB261E88C9F7777372C10CD076",
  },
  {
    "amount": "0",
    "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
  },
  {
    "amount": "0",
    "denom": "uosmo",
  },
]
`;

exports[`Test pool calculations calcMaxCoinsForPool pool#604 1`] = `
[
  {
    "amount": "67812902",
    "denom": "ibc/987C17B11ABC2B20019178ACE62929FE9840202CE79498E29FE8E5CB02B7C0A4",
  },
  {
    "amount": "5283728",
    "denom": "uosmo",
  },
]
`;

exports[`Test pool calculations convertDollarValueToCoins 0 dollar 1`] = `
[
  {
    "amount": "0",
    "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
    "displayAmount": "0",
    "symbol": "ATOM",
    "value": "0",
  },
  {
    "amount": "0",
    "denom": "uosmo",
    "displayAmount": "0",
    "symbol": "OSMO",
    "value": "0",
  },
]
`;

exports[`Test pool calculations convertDollarValueToCoins 2 token pool 1`] = `
[
  {
    "amount": "136500.13650013650014",
    "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
    "displayAmount": "0.13650013650013650014",
    "symbol": "ATOM",
    "value": "5",
  },
  {
    "amount": "549450.54945054945055",
    "denom": "uosmo",
    "displayAmount": "0.54945054945054945055",
    "symbol": "OSMO",
    "value": "5",
  },
]
`;

exports[`Test pool calculations convertDollarValueToCoins 3 token pool 1`] = `
[
  {
    "amount": "7071428.57142857142857",
    "denom": "ibc/1DCC8A6CB5689018431323953344A9F6CC4D0BFB261E88C9F7777372C10CD076",
    "displayAmount": "7.07142857142857142857",
    "symbol": "REGEN",
    "value": "9.9",
  },
  {
    "amount": "1365.001365001365",
    "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
    "displayAmount": "0.001365001365001365",
    "symbol": "ATOM",
    "value": "0.05",
  },
  {
    "amount": "5494.50549450549451",
    "denom": "uosmo",
    "displayAmount": "0.00549450549450549451",
    "symbol": "OSMO",
    "value": "0.05",
  },
]
`;

exports[`Test pool calculations convertDollarValueToCoins 4 token pool 1`] = `
[
  {
    "amount": "1785714.28571428571429",
    "denom": "ibc/1DCC8A6CB5689018431323953344A9F6CC4D0BFB261E88C9F7777372C10CD076",
    "displayAmount": "1.78571428571428571429",
    "symbol": "REGEN",
    "value": "2.5",
  },
  {
    "amount": "68250.06825006825007",
    "denom": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
    "displayAmount": "0.06825006825006825007",
    "symbol": "ATOM",
    "value": "2.5",
  },
  {
    "amount": "642673.52185089974293",
    "denom": "ibc/A0CC0CF735BFB30E730C70019D4218A1244FF383503FF7579C9201AB93CA9293",
    "displayAmount": "0.64267352185089974293",
    "symbol": "XPRT",
    "value": "2.5",
  },
  {
    "amount": "274725.27472527472527",
    "denom": "uosmo",
    "displayAmount": "0.27472527472527472527",
    "symbol": "OSMO",
    "value": "2.5",
  },
]
`;

exports[`Test pool calculations makePoolPairs 1`] = `
[
  {
    "baseAddress": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
    "baseName": "atom",
    "baseSymbol": "ATOM",
    "poolAddress": "osmo1mw0ac6rwlp5r8wapwk3zs6g29h8fcscxqakdzw9emkne6c8wjp9q0t3v8t",
    "poolId": "1",
    "quoteAddress": "uosmo",
    "quoteName": "osmo",
    "quoteSymbol": "OSMO",
  },
  {
    "baseAddress": "uion",
    "baseName": "ion",
    "baseSymbol": "ION",
    "poolAddress": "osmo1500hy75krs9e8t50aav6fahk8sxhajn9ctp40qwvvn8tcprkk6wszun4a5",
    "poolId": "2",
    "quoteAddress": "uosmo",
    "quoteName": "osmo",
    "quoteSymbol": "OSMO",
  },
  {
    "baseAddress": "ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4",
    "baseName": "akt",
    "baseSymbol": "AKT",
    "poolAddress": "osmo1c9gj5nwxhuh2gz7wwg4r8e8tw8v7ggy9lh2hu7kkdgh0t450754qh9cpvd",
    "poolId": "3",
    "quoteAddress": "uosmo",
    "quoteName": "osmo",
    "quoteSymbol": "OSMO",
  },
  {
    "baseAddress": "ibc/1480B8FD20AD5FCAE81EA87584D269547DD4D436843C1D20F15E00EB64743EF4",
    "baseName": "akt",
    "baseSymbol": "AKT",
    "poolAddress": "osmo1lzwv0glchfcw0fpwzdwfdsepmvluv6z6eh4qunxdml33sj06q3yq7xwtde",
    "poolId": "4",
    "quoteAddress": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
    "quoteName": "atom",
    "quoteSymbol": "ATOM",
  },
  {
    "baseAddress": "ibc/9712DBB13B9631EDFA9BF61B55F1B2D290B2ADB67E3A4EB3A875F3B6081B3B84",
    "baseName": "dvpn",
    "baseSymbol": "DVPN",
    "poolAddress": "osmo1j5l9ysw5xv0uqz9uh7mcg0l5rlerqm695ec9kkg2t8rp600zv47q82eqwa",
    "poolId": "5",
    "quoteAddress": "uosmo",
    "quoteName": "osmo",
    "quoteSymbol": "OSMO",
  },
  {
    "baseAddress": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
    "baseName": "atom",
    "baseSymbol": "ATOM",
    "poolAddress": "osmo1p0rpttlp8v2hy7m82l2t9p6545788f2ac3yksgrlycke2wr4mu0qdr7ytu",
    "poolId": "6",
    "quoteAddress": "ibc/9712DBB13B9631EDFA9BF61B55F1B2D290B2ADB67E3A4EB3A875F3B6081B3B84",
    "quoteName": "dvpn",
    "quoteSymbol": "DVPN",
  },
  {
    "baseAddress": "ibc/7C4D60AA95E5A7558B0A364860979CA34B7FF8AAF255B87AF9E879374470CEC0",
    "baseName": "iris",
    "baseSymbol": "IRIS",
    "poolAddress": "osmo13jr3p5p070h4pu7sxhtldag9899sev9pwx0r2vlvpkyravpxlqssnzsuq9",
    "poolId": "7",
    "quoteAddress": "uosmo",
    "quoteName": "osmo",
    "quoteSymbol": "OSMO",
  },
  {
    "baseAddress": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
    "baseName": "atom",
    "baseSymbol": "ATOM",
    "poolAddress": "osmo1605py4r32r73csszlqpz6n5t5sgax9gjjfnm68jnd7pv6qe4l54ql0w8ku",
    "poolId": "8",
    "quoteAddress": "ibc/7C4D60AA95E5A7558B0A364860979CA34B7FF8AAF255B87AF9E879374470CEC0",
    "quoteName": "iris",
    "quoteSymbol": "IRIS",
  },
  {
    "baseAddress": "ibc/E6931F78057F7CC5DA0FD6CEF82FF39373A6E0452BF1FD76910B93292CF356C1",
    "baseName": "cro",
    "baseSymbol": "CRO",
    "poolAddress": "osmo19fm8jtzyw8ujsnsqm5rznudn8fhhkykjh4ra8rvx9lsfslw2pc2sp36h3r",
    "poolId": "9",
    "quoteAddress": "uosmo",
    "quoteName": "osmo",
    "quoteSymbol": "OSMO",
  },
  {
    "baseAddress": "ibc/27394FB092D2ECCD56123C74F36E4C1F926001CEADA9CA97EA622B25F41E5EB2",
    "baseName": "atom",
    "baseSymbol": "ATOM",
    "poolAddress": "osmo1krp38zzc3zz5as9ndqkyskhkzv6x9e30ckcq5g4lcsu5wpwcqy0sa3dea2",
    "poolId": "10",
    "quoteAddress": "ibc/E6931F78057F7CC5DA0FD6CEF82FF39373A6E0452BF1FD76910B93292CF356C1",
    "quoteName": "cro",
    "quoteSymbol": "CRO",
  },
]
`;
